Readme.md
# babel-plugin-sass-vars [![npm](https://img.shields.io/npm/v/babel-plugin-sass-vars.svg)](https://www.npmjs.com/package/babel-plugin-sass-vars)

Extract [sass/scss](https://sass-lang.com) variables at build time.

```javascript
import styles from "style.scss";
import {primaryColor} from "style.scss";
```

## Features

- Import all global sass variables, across all `@import`ed files
- All sass functions, variable substitutions, etc. are evaluated by the official
  [node-sass](https://github.com/sass/node-sass) compiler
- Both sass and scss are supported and can be intermingled to the extent [supported by
  node-sass](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#syntax)

## Install

This package is available on npm and can be installed with

```
npm install -D babel-plugin-sass-vars
```

Note that the package is a [babel](https://babeljs.io) plugin and must be used as part of
a babel build step.

## Babel configuration

This plugin can be activated by adding it to the `plugins` array within babel's
[config](https://babeljs.io/docs/usage/babelrc/). For example,

```javascript
{
    "plugins": [
        ["babel-plugin-sass-vars", {
            "sassCase": // ...
            "outputCase": // ...
        }]
    ]
}
```

## Usage

The plugin transforms certain `import` and `require` statements and in general has the
following behavior:

- The target filename must contain an `.scss` or `.sass` extension to be considered
- All sass values are stringified without further processing, so for example `$var: 42`
  becomes `var: "42"`, `$var: 42px` becomes `var: "42px"`, and `$var: "hello"` becomes
  `var: "hello"`

### Default import

To extract all global variables into a single object, just import the "default export" of
the stylesheet:
```javascript
import STYLE from "./styles.scss";
```
This import statement will then be replaced with an assignment like
```javascript
const STYLE = Object.freeze({
    var: "value",
    // ...
});
```
so that each variable can be accessed through `STYLE`.

### Named imports

To import just a known subset of variables, the named form can be applied as usual:
```javascript
import {myVar, otherVar as coolVar} from "./styles.scss";
```
This statement will then be replaced with assignments like
```javascript
const myVar = "...";
const coolVar = "...";
```
Note that both import forms can be used together if desired:
```javascript
import STYLE, {myVar} from "./styles.scss";
```

### CommonJS `require` import

The CommonJS import form is also supported:
```javascript
let myStyles = require("./styles.scss");
```
This will then be transformed to something like
```javascript
let myStyles = Object.freeze({
    var: "value",
    // ...
});
```

## Options

The options object for the plugin can contain the following parameters.

### `sassCase`

This option identifies the casing format of the variables within any imported stylesheets.
All named import identifiers will be converted to this case before looking them up in the
stylesheet. This means that variables written in camelCase in the javascript file can
correspond to variables written in param-case in the stylesheet.

If only the default-import form is used, this option has no effect and can be left unset.

The value should be a string corresponding to one of the top-level functions of the
[`change-case`](https://www.npmjs.com/package/change-case) package, for example
`"paramCase"` or `"camelCase"`.

### `outputCase`

This option identifies how the casing of sass variables should be transformed when they're
exported to javascript via the default-import object or a named import variable. Note that
when using the named-import form, the named identifiers must also be in this case style.

Its value should be given in the same format as `sassCase`.

## Examples

### Basic examples

These examples assume a stylesheet with the contents
```sass
$primary-color: turqoise;
$secondary-color: salmon;
$alt-color: $primary-color;
$main-padding: 32;
$main-border: 1px solid #000;
```
is stored at the location `../styles/app.scss` relative to the javascript files.

Then with the following plugin options,
```javascript
{
    sassCase: "paramCase",
    outputCase: "camelCase",
}
```

the following input
```javascript
import styles from "../styles/app.scss"
import {primaryColor, secondaryColor as SECONDARY_COLOR} from "../styles/app.scss";
```
would be transformed to something like this output:
```javascript
const styles = Object.freeze({
    primaryColor: "turqoise",
    secondaryColor: "salmon",
    altColor: "turqoise",
    mainPadding: "32",
    mainBorder: "1px solid #000",
});

const primaryColor = "turqoise";

const SECONDARY_COLOR = "salmon";
```

Alternatively, if the babel options were set to output in constant case:
```javascript
{
    sassCase: "paramCase",
    outputCase: "constantCase",
}
```
then the input file
```javascript
import styles from "../styles/app.scss"
import {PRIMARY_COLOR, SECONDARY_COLOR as secondary} from "../styles/app.scss";
```
would be transformed to something like
```javascript
const styles = Object.freeze({
    PRIMARY_COLOR: "turqoise",
    SECONDARY_COLOR: "salmon",
    ALT_COLOR: "turqoise",
    MAIN_PADDING: "32",
    MAIN_BORDER: "1px solid #000",
});

const PRIMARY_COLOR = "turqoise";

const secondary = "salmon";
```

### CSS framework example

This example shows how the plugin can be applied to a project that uses a large CSS
framework. In this case, we'll consider a stylesheet that pulls in
[bulma](https://bulma.io):
```sass
$primary: aqua;

@import "../node_modules/bulma/bulma.sass";
```
Since bulma uses param-case for its variables, we set `sassCase` to `paramCase`, and
arbitrarily choose `camelCase` for `outputCase`:
```javascript
{
    sassCase: "paramCase",
    outputCase: "camelCase",
}
```
With this configuration, the following input file
```javascript
import styles, {primary, info, familyMonospace} from "../styles/app.scss";
```
would be transformed to something like the following:
```javascript
const styles = Object.freeze({
  "primary": "aqua",
  "black": "#0a0a0a",
  "greyDarker": "#363636",
  "greyDark": "#4a4a4a",
  "grey": "#7a7a7a",
  "greyLight": "#b5b5b5",
  "greyLighter": "#dbdbdb",
  "familyMonospace": "monospace",
  "renderMode": "optimizeLegibility",
  "light": "whitesmoke",
  "dark": "#363636",
  "orangeInvert": "#fff",
  "info": "#3273dc",
  // All other bulma variables...
});

const primary = "aqua";

const info = "#3273dc";

const familyMonospace = "monospace";
```

## How it works

This plugin is pretty much a hack, but there doesn't seem to be any existing interface
provided by node-sass for getting at variable assignments. As such, the following
procedure is used to extract these variables indirectly:

1. Create a set of desired variables:
    - If the default-import form is used, we need to find all global variables that could
      be accessed. To do so, the root stylesheet and any further imported stylesheets are
      traversed and parsed (using [gonzales-pe](https://github.com/tonyganch/gonzales-pe))
      to extract variable names defined at the top level.
    - If the named-import form is used, then the given names are simply transformed to
      sass case and used as the set.
2. Generate a stylesheet that `@import`s the root stylesheet and assigns each desired
   sass variable to an associated "unique" [custom property](https://developer.mozilla.org/en-US/docs/Web/CSS/--*).
3. Transform this generated stylesheet to a CSS stylesheet using node-sass, which performs
   function evaluation, variable substitution, and all its other processing and assigns
   each custom property the final value of each variable.
4. Parse this CSS stylesheet (using gonzales-pe), extract each custom property, and
   associate the value with each corresponding sass variable.


Changelog.md
## v0.2.1 - 2018-03-12

- Fix an example in the readme.

## v0.2.0 - 2018-03-12

### Fixes

- Now works with node-sass@4.8.0/libsass@3.5.0, which fixes "Please check validity of the
  block" errors.

### Breaking changes

- Quoted SASS vars no longer contain unnecessary surrounding quotes in their javascript
  representation. For example, the variable `$myvar: "a b c"` used to be output as `const
  myvar = '"a b c"'`, but is now output as `const myvar = 'a b c'`.

## v0.1.1 - 2018-02-28

- Initial release.


test/index.js
import {assert} from "chai";
import path from "path";
import {camelCase} from "change-case";

import {fileSyntax, genCssProps, VarNames, SassVars} from "../src";

test("fileSyntax", () => {
    assert.deepEqual(fileSyntax("abc.scss"), ["abc.scss", "scss"]);
    assert.deepEqual(fileSyntax("abc.sass"), ["abc.sass", "sass"]);

    assert.deepEqual(fileSyntax(path.join(__dirname, "testSass")),
        [path.join(__dirname, "testSass.sass"), "sass"]);
    assert.deepEqual(fileSyntax(path.join(__dirname, "testScss")),
        [path.join(__dirname, "testScss.scss"), "scss"]);
    assert.deepEqual(fileSyntax(path.join(__dirname, "testAmbig")),
        [path.join(__dirname, "testAmbig.sass"), "sass"]);

    assert.throws(() => fileSyntax("abc.html"));
});

test("VarNames", () => {
    let vars = new VarNames(path.join(__dirname, "testScss.scss")).extract();

    assert.equal(vars.size, 4);
    assert.isTrue(vars.has("scssvar"));
    assert.isTrue(vars.has("sassvar"));
    assert.isTrue(vars.has("ambigvar"));
    assert.isTrue(vars.has("mapvar"));
});

test("SassVars", () => {
    let props = genCssProps(["scssvar", "sassvar", "ambigvar", "mapvar"]);
    let vars = new SassVars(path.join(__dirname, "testScss.scss"),
                            props, camelCase).extract();

    assert.deepEqual(vars, {
        sassvar: "42",
        scssvar: "69",
        ambigvar: "green",
        mapvar: "(abc: 123)",
    });
});


test/.babel-common.js
module.exports = opts => {
    require("@babel/register")({
        cache: false,
        ignore: [/node_modules|lib/],
        presets: [
            ["@babel/preset-env", {
                targets: {node: "current"},
                useBuiltIns: "usage",
                shippedProposals: true,
            }],
        ],
        plugins: [
            ["@babel/plugin-proposal-object-rest-spread", {useBuiltIns: true}],
            ["./lib", opts],
        ],
    });
};


test/babel/testNoOpts/index.js
import {assert} from "chai";

import sty from "./testScss.scss";

import {mapvar} from "./testScss.scss";

import styles, {sassvar, scssvar} from "../../testScss.scss";

test("default import", () => {
    assert.deepEqual(sty, {
        scssvar: "69em",
        mapvar: "(abc: 123)",
        sassvar: "42",
    });
});

test("named import", () => {
    assert.equal(mapvar, "(abc: 123)");
});

test("mixed import", () => {
    assert.deepEqual(styles, {
        scssvar: "69",
        mapvar: "(abc: 123)",
        sassvar: "42",
        ambigvar: "green",
    });

    assert.equal(sassvar, "42");
    assert.equal(scssvar, "69");
});


src/index.js
import {readFileSync, existsSync} from "fs";
import {resolve, dirname, extname} from "path";

import changeCase from "change-case";
import memoize from "fast-memoize";
import {parse} from "gonzales-pe";
import {renderSync} from "node-sass";
import serialize from "babel-literal-to-ast";
import uuid from "uuid/v4";

// Matches supported extensions.
const EXT_RE = /(?:\.scss|\.sass)$/;

// Random string to make CSS properties "unique".
const RAND = uuid();

// Default babel options.
const DEFAULT_OPTS = Object.freeze({
    sassCase: null,
    outputCase: null,
});

export default function({types: t}, opts) {
    opts = Object.assign({}, DEFAULT_OPTS, opts);

    let sassCase = opts.sassCase ?
        changeCase[opts.sassCase] : sameCase;
    let outputCase = opts.outputCase ?
        changeCase[opts.outputCase] : sameCase;

    let lookup = new VarLookup(sassCase, outputCase);

    let visitor = {
        ImportDeclaration(astPath, {file}) {
            let importPath = astPath.node.source.value;

            if (!EXT_RE.test(importPath)) {
                return;
            }

            let absPath = resolveAbsPath(file.opts.filename, importPath);
            let specs = astPath.node.specifiers;
            let replace = [];

            if (t.isImportDefaultSpecifier(specs[0])) {
                let vars = lookup.extractAllVars(absPath);

                replace.push(t.variableDeclaration("const", [
                    t.variableDeclarator(specs[0].local, createVarsObject(t, vars)),
                ]));

                specs.shift();
            }

            if (specs.length > 0) {
                let names = specs.map(node => node.imported.name);
                let vars = lookup.extractVars(absPath, names);

                specs.forEach(node => {
                    replace.push(t.variableDeclaration("const", [
                        t.variableDeclarator(
                            node.local,
                            serialize(vars[node.imported.name])
                        )
                    ]));
                });
            }

            astPath.replaceWithMultiple(replace);
        },
        CallExpression(astPath, {file}) {
            let calleeName = astPath.node.callee.name;

            if (calleeName !== "require") {
                return;
            }

            let args = astPath.node.arguments;

            if (args.length !== 1 || !t.isStringLiteral(args[0])) {
                return;
            }

            let importPath = args[0].value;

            if (!EXT_RE.test(importPath)) {
                return;
            }

            let absPath = resolveAbsPath(file.opts.filename, importPath);
            let vars = lookup.extractAllVars(absPath);

            astPath.replaceWith(createVarsObject(t, vars));
        }
    };

    return {visitor};
}

class VarLookup {
    constructor(sassCase, outputCase) {
        this._sassCase = sassCase;
        this._outputCase = outputCase;

        // Maps filenames to a dictionary of known variable name/value pairs from that
        // file.
        //
        // Each variable name is in `outputCase`.
        this._fileVars = {};

        this.extractAllVars = memoize(this._extractAllVars.bind(this));
    }

    _extractAllVars(absPath) {
        return this._extractVars(absPath, this._extractNames(absPath));
    }

    _extractNames(absPath) {
        return genCssProps(new VarNames(absPath).extract());
    }

    extractVars(absPath, names) {
        let knownVars = this._lookupVars(absPath);

        let unknownNames = [...names].filter(name => (
            !knownVars.hasOwnProperty(name)
        ));

        if (unknownNames.length === 0) {
            return knownVars;
        }

        let props = genCssProps(unknownNames.map(this._sassCase));
        let vars = this._extractVars(absPath, props);

        if (!unknownNames.every(name => vars.hasOwnProperty(name))) {
            throw new Error("import names must be in the same case as `outputCase`");
        }

        return vars;
    }

    _extractVars(absPath, props) {
        let vars = new SassVars(absPath, props, this._outputCase).extract();
        let knownVars = this._lookupVars(absPath);

        Object.assign(knownVars, vars);

        return knownVars;
    }

    _lookupVars(absPath) {
        if (!this._fileVars.hasOwnProperty(absPath)) {
            this._fileVars[absPath] = {};
        }

        return this._fileVars[absPath];
    }
}

class SassVars {
    constructor(initPath, nameProps, caseFn) {
        this._initPath = initPath;
        this._namesProps = nameProps;
        this._caseFn = caseFn;
        this._vars = {};
    }

    extract() {
        let result = renderSync({
            data: this._buildSassData(),
            outputStyle: "expanded",
        });

        let output = result.css.toString();
        this._processOutput(output);

        return this._vars;
    }

    _processOutput(buf) {
        parse(buf).traverse((node, idx, parent) => {
            if (node.type !== "customProperty") {
                return;
            }

            let ident = node.content[0].content;
            let name = this._namesProps[ident];

            if (!name) {
                return;
            }

            let str = parent.content[parent.content.length - 1].toString();
            let val = str.substring(1, str.length - 1);

            this._vars[this._caseFn(name)] = val;
        });
    }

    _buildSassData() {
        let str = [];

        str.push(`@import "${this._initPath}";\n`);
        str.push(`#vars_${RAND} {\n`);

        Object.entries(this._namesProps).forEach(([prop, name]) => {
            str.push(`--${prop}: "#{$${name}}";\n`);
        });

        str.push("}");

        return str.join("");
    }

}

class VarNames {
    constructor(absPath) {
        this._initPath = absPath;
        this._names = new Set();
    }

    extract() {
        this._extractNames(this._initPath);
        return this._names;
    }

    _extractNames(filePath) {
        let [fullPath, syntax] = fileSyntax(filePath);
        let contents = readFileSync(fullPath, {encoding: "utf8"});

        parse(contents, {syntax}).forEach(node => {
            switch (node.type) {
            case "atrule":
                this._processImport(node, fullPath);
            break;
            case "declaration":
                this._processDecl(node);
            break;
            }
        });
    }

    _processDecl(node) {
        let name = this._getName(node);
        this._names.add(name);
    }

    _getName(node) {
        let propNode = node.content[0];
        let varNode = propNode.content[0];
        let identNode = varNode.content[0];

        return identNode.content;
    }

    _processImport(node, filePath) {
        let kwNode = node.first("atkeyword");

        if (kwNode === null) {
            return;
        }

        let identNode = kwNode.content[0];

        if (identNode.content !== "import") {
            return;
        }

        let pathNode = node.first("string");

        if (pathNode === null) {
            return;
        }

        let importPath = pathNode.content.slice(1, -1);
        let absPath = resolveAbsPath(filePath, importPath);

        this._extractNames(absPath);
    }
}

function createVarsObject(t, vars) {
    return t.callExpression(
        t.memberExpression(
            t.identifier("Object"),
            t.identifier("freeze")
        ),
        [serialize(vars)]
    );
}

function resolveAbsPath(requiringPath, importPath) {
    return resolve(dirname(requiringPath), importPath);
}

function fileSyntax(path) {
    let ext = extname(path);

    switch (ext) {
    case ".sass": return [path, "sass"];
    case ".scss": return [path, "scss"];
    case "":
        for (let testExt of ["sass", "scss"]) {
            let testPath = `${path}.${testExt}`;

            if (existsSync(testPath)) {
                return [testPath, testExt];
            }
        }

        throw new Error(`invalid import ${path}`);
    }

    throw new Error(`unknown file extension ${ext}`);
}

function genCssProps(names) {
    let props = {};

    names.forEach(name => {
        props[`${name}_${RAND}`] = name;
    });

    return props;
}

function sameCase(x) {
    return x;
}

if (process.env.NODE_TEST) {
    let {assert} = require("chai");
    let path = require("path");

    test("fileSyntax", () => {
        assert.deepEqual(fileSyntax("abc.scss"), ["abc.scss", "scss"]);
        assert.deepEqual(fileSyntax("abc.sass"), ["abc.sass", "sass"]);

        assert.deepEqual(fileSyntax(path.join(__dirname, "../test/testSass")),
            [path.join(__dirname, "../test/testSass.sass"), "sass"]);
        assert.deepEqual(fileSyntax(path.join(__dirname, "../test/testScss")),
            [path.join(__dirname, "../test/testScss.scss"), "scss"]);
        assert.deepEqual(fileSyntax(path.join(__dirname, "../test/testAmbig")),
            [path.join(__dirname, "../test/testAmbig.sass"), "sass"]);

        assert.throws(() => fileSyntax("abc.html"));
    });

    test("VarNames", () => {
        let vars = new VarNames(path.join(__dirname, "../test/testScss.scss")).extract();

        assert.equal(vars.size, 4);
        assert.isTrue(vars.has("scssvar"));
        assert.isTrue(vars.has("sassvar"));
        assert.isTrue(vars.has("ambigvar"));
        assert.isTrue(vars.has("mapvar"));
    });

    test("SassVars", () => {
        let props = genCssProps(["scssvar", "sassvar", "ambigvar", "mapvar"]);
        let vars = new SassVars(path.join(__dirname, "../test/testScss.scss"),
                                props, changeCase.camelCase).extract();

        assert.deepEqual(vars, {
            sassvar: "42",
            scssvar: "69",
            ambigvar: "green",
            mapvar: "(abc: 123)",
        });
    });

    test("VarLookup", () => {
        let lookup = new VarLookup(changeCase.paramCase, changeCase.constantCase);

        let vars = lookup.extractAllVars(path.join(__dirname, "../test/testScss.scss"));
        assert.deepEqual(vars, {
            SASSVAR: "42",
            SCSSVAR: "69",
            AMBIGVAR: "green",
            MAPVAR: "(abc: 123)",
        });

        vars = lookup.extractVars(path.join(__dirname, "../test/testScss.scss"),
            ["SASSVAR", "MAPVAR"]);
        assert.deepEqual(vars, {
            SASSVAR: "42",
            SCSSVAR: "69",
            AMBIGVAR: "green",
            MAPVAR: "(abc: 123)",
        });

        lookup = new VarLookup(changeCase.paramCase, changeCase.constantCase);

        vars = lookup.extractVars(path.join(__dirname, "../test/testVars.scss"),
            ["MY_VAR", "ANOTHER_COOL_VAR"]);
        assert.deepEqual(vars, {
            MY_VAR: "42",
            ANOTHER_COOL_VAR: "69",
        });

        lookup = new VarLookup(changeCase.paramCase, changeCase.camelCase);

        vars = lookup.extractVars(path.join(__dirname, "../test/testVars.scss"),
            ["myVar", "sassvar"]);
        assert.deepEqual(vars, {
            myVar: "42",
            sassvar: "42",
        });
    });
}
